Add scheduler.yield() (#35960)
mdn/content: files/en-us/web/api/prioritized_task_scheduling_api/index.md

The Prioritized Task Scheduling API is available in both window and worker threads using the `scheduler` property on the global object.

> The main API methods are {{domxref('scheduler.postTask()')}} and {{domxref('scheduler.yield()')}}. `scheduler.postTask()` takes a callback function ("the task") and returns a promise that resolves with the return value of the function, or rejects with an error. `scheduler.yield()` turns any async function into a task by yielding the main thread to the browser for other work, with execution continuing when the returned promise is resolved.

Reviewer: Worth calling out some more differences here?

- That `postTask` has more options for setting priority and/or aborting the task
- But that `scheduler.yield` is simpler to use as can be added without wrapping code in a function.

> The main API methods are {{domxref('scheduler.postTask()')}} and {{domxref('scheduler.yield()')}}. `scheduler.postTask()` takes a callback function ("the task") and returns a promise that resolves with the return value of the function, or rejects with an error. `scheduler.yield()` turns any async function into a task by yielding the main thread to the browser for other work, with execution continuing when the returned promise is resolved.

Reviewer 2: ```suggestion
The main API methods are {{domxref('scheduler.postTask()')}} and {{domxref('scheduler.yield()')}}. `scheduler.postTask()` takes a callback function (the task) and returns a promise that resolves with the return value of the function, or rejects with an error. `scheduler.yield()` turns any async function into a task by yielding the main thread to the browser for other work, with execution continuing when the returned promise is resolved.
```

#### `scheduler.yield()`

> To break up long-runing JavaScript so it doesn't block the main thread, a `scheduler.yield()` can be inserted to temporarily yield the main thread back to the browser, which creates a task to continue execution where it left off.

Reviewer 2: ```suggestion
To break up long-running JavaScript tasks so they don'tt block the main thread, a call to `scheduler.yield()` can be inserted to temporarily yield the main thread back to the browser, which creates a task to continue execution where it left off.
```

```js
async function slowTask() {
  firstHalfOfWork();
  await scheduler.yield();
  secondHalfOfWork();
}
```

> `scheduler.yield()` returns a promise that can be awaited for when execution can continue. This allows work that belongs in the same function to stay in the same function but without blocking the main thread the entire time the function runs.

Reviewer 2: ```suggestion
`scheduler.yield()` returns a promise that can be awaited for when execution can continue. This allows work that belongs in the same function to stay in the same function, but without blocking the main thread the entire time the function runs.
```

> `scheduler.yield()` takes no arguments. The task that triggers its continuation has a default [`user-visible`](#user-visible) priority, but if a `scheduler.yield()` is run within a `scheduler.postTask()` task, it will [inherit the priority of the surrounding task](/en-US/docs/Web/API/Scheduler/yield#inheriting_task_priorities).

Reviewer 2: ```suggestion
`scheduler.yield()` takes no arguments. The task that triggers its continuation has a default [`user-visible`](#user-visible) priority, but if `scheduler.yield()` is called within a `scheduler.postTask()` callback, it will [inherit the priority of the surrounding task](/en-US/docs/Web/API/Scheduler/yield#inheriting_task_priorities).
```

#### `scheduler.postTask()`

> `scheduler.postTask()` creates a task with default priority [`user-visible`](#user-visible) that has a fixed priority and cannot be aborted.

Reviewer: Maybe:

```suggestion
`scheduler.postTask()` creates a task with an optional priority (defaulted to [`user-visible`](#user-visible)) that cannot be aborted.
```

And can it not be aborted? I thought that was one of the benefits of `postTask` over `yield` (see my comment above). What's https://developer.mozilla.org//en-US/docs/Web/API/Scheduler/postTask#aborting_tasks all about?

Or does this mean the default call can't be aborted (since it doesn't provide a `signal` argument). If so this should be reworded (appreciate you didn't write the original, but still let's fix this while in here).

The Prioritized Task Scheduling API is available in both window and worker threads using the `scheduler` property on the global object.

The main API methods are {{domxref('scheduler.postTask()')}} and {{domxref('scheduler.yield()')}}. `scheduler.postTask()` takes a callback function (the task) and returns a promise that resolves with the return value of the function, or rejects with an error. `scheduler.yield()` turns any async function into a task by yielding the main thread to the browser for other work, with execution continuing when the returned promise is resolved.

> `scheduler.postTask()` is more configurable, like explicitly setting its task priority or `AbortSignal` to cancel the task. `scheduler.yield()` is simpler and can be `await`ed in any async function without having to provide the followup task in another function.

Reviewer 3: ```suggestion
The two methods have similar functionality but different levels of control. `scheduler.postTask()` is more configurable — for example, it allows task priority to be explicitly set and task cancellation via an [`AbortSignal`](/en-US/docs/Web/API/AbortSignal). `scheduler.yield()` on the other hand is simpler and can be `await`ed in any `async` function without having to provide a followup task in another function.
```

#### `scheduler.yield()`

To break up long-running JavaScript tasks so they don't block the main thread, a call to `scheduler.yield()` can be inserted to temporarily yield the main thread back to the browser, which creates a task to continue execution where it left off.

```js
async function slowTask() {
  firstHalfOfWork();
  await scheduler.yield();
  secondHalfOfWork();
}
```

`scheduler.yield()` returns a promise that can be awaited for when execution can continue. This allows work that belongs in the same function to stay in the same function, but without blocking the main thread the entire time the function runs.

`scheduler.yield()` takes no arguments. The task that triggers its continuation has a default [`user-visible`](#user-visible) priority, but if `scheduler.yield()` is called within a `scheduler.postTask()` callback, it will [inherit the priority of the surrounding task](/en-US/docs/Web/API/Scheduler/yield#inheriting_task_priorities).

#### `scheduler.postTask()`

> `scheduler.postTask()` called with no arguments creates a task with a default [`user-visible`](#user-visible) priority that cannot be aborted or have its priority changed.

Author: Trying to clarify that this is just when it's called with no arguments:
```suggestion
When `scheduler.postTask()` is called with no arguments, it creates a task with a default [`user-visible`](#user-visible) priority that cannot be aborted or have its priority changed.
```
